File: spring-boot-starter-breuninger-jobs/src/main/java/com/breuninger/boot/jobs/repository/JobRepository.java
package com.breuninger.boot.jobs.repository;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Optional;

import com.breuninger.boot.jobs.domain.JobInfo;
import com.breuninger.boot.jobs.domain.JobInfo.JobStatus;
import com.breuninger.boot.jobs.domain.JobMessage;

public interface JobRepository {

  Optional<JobInfo> findOne(String jobId);

  List<JobInfo> findLatest(int maxCount);

  List<JobInfo> findLatestJobsDistinct();

  List<JobInfo> findLatestBy(String type, int maxCount);

  List<JobInfo> findRunningWithoutUpdateSince(OffsetDateTime timeOffset);

  List<JobInfo> findAll();

  List<JobInfo> findAllJobInfoWithoutMessages();

  List<JobInfo> findByType(String jobType);

  JobInfo createOrUpdate(JobInfo job);

  void removeIfStopped(String jobId);

  JobStatus findStatus(String jobId);

  void appendMessage(String jobId, JobMessage jobMessage);

  void setJobStatus(String jobId, JobStatus jobStatus);

  void setLastUpdate(String jobId, OffsetDateTime lastUpdate);

  long size();

  void deleteAll();
}

File: jobs/src/main/kotlin/com/breuninger/boot/jobs/repository/JobRepository.kt
package com.breuninger.boot.jobs.repository

import com.breuninger.boot.jobs.domain.Job
import com.breuninger.boot.jobs.domain.JobBlockedException
import com.breuninger.boot.jobs.domain.JobExecutionId
import com.breuninger.boot.jobs.domain.JobId

interface JobRepository {

  fun findOne(jobId: JobId): Job?

  fun findRunning(jobIds: Set<JobId>): Job?

  fun findAll(): List<Job>

  fun insert(job: Job)

  fun updateDisableState(jobId: JobId, job: Job): Job?

  fun acquireRunLock(jobId: JobId, jobExecutionId: JobExecutionId): Job?

  @Throws(JobBlockedException::class)
  fun releaseRunLock(jobId: JobId, jobExecutionId: JobExecutionId): Unit?

  fun findState(jobId: JobId, key: String): String?

  fun updateState(jobId: JobId, key: String, value: String?): Unit?
}
